articles/modules/ROOT/pages/using-subqueries-to-avoid-the-eager.adoc

```diff
@@ -0,0 +1,222 @@
+and not on the main branch of execution, which is the direct line of operators from the top-leftmost operator (which may not be at the top of the plan, so check carefully) to the last operator at the bottom.
+
+== Nested subqueries for additional scoping
+
+For a more complex query, a single subquery usage may not be enough to properly reign in the eager behavior.
+
+That is, when the Eager is scoped behind a subquery, it means each individual subquery execution behaves eagerly, and that's usually enough to make the impact minimal.
+But when an individual subquery execution can generate a ton of rows (such as additional MATCHes) such that the Eager still retains a negative impact,
+it may be necessary to use another subquery to scope the Eager down to yet another level.
+
+It is important that nested subqueries are applied such that they are still logically correct and produce correct results with respect to the scoping.
+
+For example, if you need to aggregate, be aware that if you aggregate within a subquery, you will be performing aggregation per subquery execution, that is its scope.
+If your aggregation needs to aggregate beyond the scope of per subquery execution, then it belongs outside of the subquery so it has visilibity for the wider scope.
+
+More on usage of aggregations and subqueries can be found here:
+
+https://support.neo4j.com/hc/en-us/articles/4403024564243-Using-Subqueries-to-Control-the-Scope-of-Aggregations
+
+== Using APOC procs as subqueries
+
+If you aren't running Neo4j 4.1 or higher, you can make use of some procs in APOC to act as subqueries for a similar effect.
+
+Notably, `apoc.cypher.run()` for read subqueries, and `apoc.cypher.doIt()` when it needs to write to the graph.
+
+Here's another similar query that results in an Eager operator in the plan:
+
+[source,cypher]
+----
+EXPLAIN
+UNWIND [1,2,3,4,5] as id
+MERGE (c:Customer {id: id})
+MERGE (e:Employee {id: c.id*10})
+ON CREATE SET e:Customer
+WITH c, e
+MERGE (e)-[r:DEDICATED_TO]->(c)
+----
+
+In this one, we conditionally add the :Customer label to the :Employee node.
+Now the two MERGEs might interfere with each other across rows such that an Eager will be planned to preserve Cypher semantics.
+
+We can apply `apoc.cypher.doIt()` as a subquery, using it similar as we would a native subquery:
+
+[source,cypher]
+----
+EXPLAIN
+UNWIND [1,2,3,4,5] as id
+MERGE (c:Customer {id: id})
+WITH c
+CALL apoc.cypher.doIt("
+	MERGE (e:Employee {id: c.id*10})
+	ON CREATE SET e:Customer",
+    {c:c}) YIELD value
+WITH c, value.e as e
+MERGE (e)-[r:DEDICATED_TO]->(c)
+----
+
+Just like before, isolating the scope with a subquery prevents the planner from adding the Eager, it vanishes from the query plan.
+
+Be aware, however, that usage of APOC procedures that execute a dynamic query like this require overhead to parse, compile, and execute the query, a cost that you do not have to pay when using native Cypher subqueries.
```

Comment:
and there can be a potential of Cypher injection when executing subqueries as strings